File: java/src/main/java/io/aurigraph/dlt/sdk/namespace/GovernanceApi.java
package io.aurigraph.dlt.sdk.namespace;

import com.fasterxml.jackson.core.type.TypeReference;
import io.aurigraph.dlt.sdk.AurigraphClient;
import io.aurigraph.dlt.sdk.dto.Proposal;
import io.aurigraph.dlt.sdk.dto.TreasuryStats;
import io.aurigraph.dlt.sdk.dto.VoteReceipt;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

/**
 * Namespace for /api/v11/governance endpoints — on-chain governance.
 */
public class GovernanceApi {

    private final AurigraphClient client;

    public GovernanceApi(AurigraphClient client) {
        this.client = client;
    }

    /**
     * List all governance proposals.
     */
    public List<Proposal> listProposals() {
        return client.get("/governance/proposals", new TypeReference<List<Proposal>>() {});
    }

    /**
     * Cast a vote on a governance proposal.
     */
    public VoteReceipt vote(String proposalId, boolean approve) {
        String enc = URLEncoder.encode(proposalId, StandardCharsets.UTF_8);
        return client.post("/governance/proposals/" + enc + "/vote",
                Map.of("approve", approve), VoteReceipt.class);
    }

    /**
     * Get governance treasury statistics.
     */
    public TreasuryStats getTreasury() {
        return client.get("/governance/treasury", TreasuryStats.class);
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/apis/GovernanceApi.kt
package io.aurigraph.sdk.apis

import io.aurigraph.sdk.AurigraphClient
import io.aurigraph.sdk.models.Proposal
import io.aurigraph.sdk.models.TreasuryStats
import io.aurigraph.sdk.models.VoteReceipt
import io.aurigraph.sdk.models.VoteRequest

/**
 * On-chain governance -- proposals, voting, and treasury.
 */
class GovernanceApi(private val client: AurigraphClient) {

    /** List all governance proposals. */
    suspend fun listProposals(): List<Proposal> = client.get("/governance/proposals")

    /** Get a specific governance proposal by ID. */
    suspend fun getProposal(proposalId: String): Proposal =
        client.get("/governance/proposals/$proposalId")

    /** Cast a vote on a governance proposal. */
    suspend fun vote(proposalId: String, approve: Boolean): VoteReceipt =
        client.post("/governance/proposals/$proposalId/vote", VoteRequest(approve))

    /** Get governance treasury statistics. */
    suspend fun getTreasury(): TreasuryStats = client.get("/governance/treasury")
}
